saivpp/src/SwitchStateBaseAcl.cpp

```diff
@@ -461,202 +604,458 @@ sai_status_t acl_priority_attr_get (
+                    // For stats we need to find vpp rule index from acl_entry_counter (ace_counter)
+                    m_ace_cntr_info_map[ace_cntr_oid] = { tbl_oid, ace.ace_oid, acl_swindex, index };
+                }
+            }
+            index++;
+        }
+    } else {
+        SWSS_LOG_ERROR("Vpp acl add replace failed, status: %d", status);
+        return SAI_STATUS_FAILURE;
+    }
 
-	index = 0;
+    SWSS_LOG_INFO("ACL table %s %s ", tbl_sid.c_str(),
+            acl_replace ? "replaced" : "added");
 
-	for (auto ace: ordered_aces) {
-	    p_ace = &aces[ace.index];
-	    const sai_attribute_t *attr;
-	    sai_object_id_t ace_cntr_oid;
+    return SAI_STATUS_SUCCESS;
+}
 
-	    for (uint32_t i = 0; i < p_ace->attrs_count; i++) {
-		attr = &p_ace->attrs[i];
+sai_status_t SwitchStateBase::tunterm_acl_bindunbind(sai_object_id_t tbl_oid, bool is_add, std::string hwif_name)
+{
+    auto tunterm_idx_it = m_tunterm_acl_swindex_map.find(tbl_oid);
+    if (tunterm_idx_it != m_tunterm_acl_swindex_map.end()) {
+        auto tunterm_idx = tunterm_idx_it->second;
+        vpp_tunterm_acl_interface_add_del(tunterm_idx, is_add, hwif_name.c_str());
+    }
+    return SAI_STATUS_SUCCESS;
+}
 
-		if (attr->id == SAI_ACL_ENTRY_ATTR_ACTION_COUNTER) {
-		    ace_cntr_oid = attr->value.aclaction.parameter.oid;
+sai_status_t SwitchStateBase::tunterm_acl_add_replace(vpp_tunterm_acl_t *acl, sai_object_id_t tbl_oid)
+{
+    sai_status_t status = SAI_STATUS_SUCCESS;
+    uint32_t     tunterm_acl_swindex = 0;
+    bool         do_port_bind = false;
+    auto         tbl_sid = sai_serialize_object_id(tbl_oid);
+
+    auto tunterm_idx_it = m_tunterm_acl_swindex_map.find(tbl_oid);
+    if ((tunterm_idx_it == m_tunterm_acl_swindex_map.end()) && (acl != NULL)) {
+        tunterm_acl_swindex = ~tunterm_acl_swindex;
+        do_port_bind = true;
+    } else if (tunterm_idx_it != m_tunterm_acl_swindex_map.end() && (acl == NULL)) {
+        status = tunterm_acl_delete(tbl_oid, false);
+        return status;
+    } else if (tunterm_idx_it != m_tunterm_acl_swindex_map.end()) {
+        tunterm_acl_swindex = tunterm_idx_it->second;
+    } else {
+        return status;
+    }
 
-		    auto ace_it = m_ace_cntr_info_map.find(ace_cntr_oid);
+    std::list<std::string> hwif_names;
+    SWSS_LOG_INFO("Adding tunterm acl rules, tunterm_acl index %u", tunterm_acl_swindex);
 
-		    if (ace_it != m_ace_cntr_info_map.end()) {
-			m_ace_cntr_info_map.erase(ace_it);
-		    }
-	
-		    // For stats we need to find vpp rule index from acl_entry_counter (ace_counter)
-		    m_ace_cntr_info_map[ace_cntr_oid] = { tbl_oid, ace.ace_oid, acl_swindex, index };
-		}
-	    }
-	    index++;
-	}
+    status = vpp_tunterm_acl_add_replace(&tunterm_acl_swindex, acl->count, acl);
+
+    if (status == SAI_STATUS_SUCCESS) {
+        m_tunterm_acl_swindex_map[tbl_oid] = tunterm_acl_swindex;
+    } else {
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (do_port_bind) {
```

Comment:
Optional: add a comment here explaining the intent of this block, ie. "enable tunterm ACL on ports to which the table is already bound"

Reply:
Updated